### PR TITLE
Always use object store to create PK table

### LINKED ISSUE #5231: Fix PrimaryKeyTests#removingPrimaryKeyRemovesConstraint_typeSetters
This unit test surfaces a problem with our schema cache:
```
   /**
     * This test surfaces a bunch of problems, most of them seem to be around caching of the schema
     * during a transaction
     *
     * 1) Removing the primary key do not invalidate the cache in RealmSchema and those cached
     *    are ImmutableRealmObjectSchema so do not change when the primary key is removed.
     *
     * 2) Addding `schema.refresh()` to RealmObjectSchema.removePrimaryKey()` causes
     *    RealmPrimaryKeyConstraintException anyway. Unclear why.
     */
```

### PR DESCRIPTION
**Pull request by beeender**
- Add OsObjectStore class to wrap methods in ObjectStore.hpp.
- Use ObjectStore to create meta tables.
- Use ObjectStore to get/set primary key.
- Always create meta tables when open a non-exist, non-readonly Dynamic
  Realm.
- Clean code.

Close #5231

### CODE REVIEW

File: realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java [hunk 1]
```diff
@@ -672,13 +674,10 @@ public void apply(DynamicRealmObject obj) {
...
 
         Table table = realm.getSchema().getTable(MigrationFieldTypeToInt.class);
-        assertTrue(table.hasPrimaryKey());
         assertEquals(MigrationFieldTypeToInt.DEFAULT_FIELDS_COUNT, table.getColumnCount());
-        assertEquals(MigrationFieldTypeToInt.DEFAULT_PRIMARY_INDEX, table.getPrimaryKey());
+        assertPKField(realm, MigrationFieldTypeToInt.CLASS_NAME, MigrationFieldTypeToInt.FIELD_PRIMARY,
+                MigrationFieldTypeToInt.DEFAULT_PRIMARY_INDEX);
 
-        RealmObjectSchema objectSchema = realm.getSchema().get(MigrationFieldTypeToInt.CLASS_NAME);
-        assertFalse(objectSchema.hasField(MigrationPrimaryKey.FIELD_PRIMARY));
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java]: This check no longer seem to happen

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncedRealmMigrationTests.java [hunk 2]
```diff
@@ -107,7 +108,7 @@ public void missingFields_hiddenSilently() {
                 .addField("newField", String.class);
         // A schema version has to be set otherwise Object Store will try to initialize the schema again and reach an
         // error branch. That is not a real case.
-        dynamicRealm.setVersion(0);
+        OsObjectStore.setSchemaVersion(dynamicRealm.getSharedRealm(), 0);
```
Comment (cmelchior) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncedRealmMigrationTests.java]: Why not have this implementation inside `dynamicRealm.setVersion(0)`, that seems more readable when using the `dynamicRealm`. With this change new developers need to know they have to go here to change the version number.

File: realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java (hunk 1 above)
Comment (beeender) [reply realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java]: added it to `assertPKField`

File: realm/realm-library/src/androidTest/java/io/realm/RealmAnnotationTests.java [hunk 3]
```diff
@@ -99,76 +101,6 @@ public void index() {
         assertFalse(table.hasSearchIndex(table.getColumnIndex("notIndexDate")));
     }
 
-    // Tests migrating primary key from string to long with existing data.
-    @Test
-    public void primaryKey_migration_long() {
```
[the remark this answers is not in the document]
Comment (beeender) [reply realm/realm-library/src/androidTest/java/io/realm/RealmAnnotationTests.java]: The duplicated values are tested with `addPrimaryKeyFieldModifier_duplicateValues`. the `primaryKey_migration_long` and `primaryKey_migration_string` are testing wrong things. will create two new tests to cover those.

File: realm/realm-library/src/main/cpp/io_realm_internal_OsObjectSchemaInfo.cpp [hunk 4]
```diff
@@ -109,3 +109,18 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_OsObjectSchemaInfo_nativeGetPrope
...
+JNIEXPORT jlong JNICALL Java_io_realm_internal_OsObjectSchemaInfo_nativeGetPrimaryKeyProperty(JNIEnv*, jclass,
+                                                                                              jlong native_ptr)
+{
+    TR_ENTER_PTR(native_ptr)
+
+    // no throws
+    auto& object_schema = *reinterpret_cast<ObjectSchema*>(native_ptr);
+    auto* property = object_schema.primary_key_property();
+    if (property) {
+        return reinterpret_cast<jlong>(new Property(*property));
```
Comment (kneth) [inline realm/realm-library/src/main/cpp/io_realm_internal_OsObjectSchemaInfo.cpp]: `new` can throw, right?

File: realm/realm-library/src/main/cpp/io_realm_internal_OsObjectStore.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            // Check valid column type
+            auto field_type = table->get_column_type(pk_column_ndx);
+            if (field_type != type_Int && field_type != type_String) {
+                THROW_JAVA_EXCEPTION(
+                    env, JavaExceptionDef::IllegalArgument,
+                    format("Field '%1' is not a valid primary key type.", StringData(pk_field_name_accessor)));
+            }
+
+            // Check duplicated values. The pk field must have been indexed before set as a PK.
+            if (table->get_distinct_view(pk_column_ndx).size() != table->size()) {
```
Comment (kneth) [inline realm/realm-library/src/main/cpp/io_realm_internal_OsObjectStore.cpp]: Should Object Store's `validate_primary_column_uniqueness()` be exposed to bindings?

File: realm/realm-library/src/main/cpp/io_realm_internal_OsObjectSchemaInfo.cpp (hunk 4 above)
Comment (beeender) [reply realm/realm-library/src/main/cpp/io_realm_internal_OsObjectSchemaInfo.cpp]: OK ... it throws `bad_alloc` ... will change this.

File: realm/realm-library/src/main/cpp/io_realm_internal_OsObjectStore.cpp (hunk 5 above)
Comment (beeender) [reply realm/realm-library/src/main/cpp/io_realm_internal_OsObjectStore.cpp]: I guess it is fine, it is only one line of code + we already have the column index :P

**Pull request merged**